addon/corpus.js

```diff
@@ -1,106 +1,231 @@
+                // Timeout.
+                console.error(tab.url, 'timeout');
+                clearTimeout(timer);
+                setCurrentStatus('timeout', true);
+
+                // Close the tab and process the next url.
+                browser.tabs.remove(tab.id)
+                    .then(() => {
+                        document.dispatchEvent(new CustomEvent(
+                            'fathom:next',
+                            { detail: windowId }
+                        ));
+                    })
+            }, gTimeout * 1000);
+
+            // Freeze this page.
+            document.dispatchEvent(new CustomEvent(
+                'fathom:freeze',
+                { detail: { windowId: windowId, timer: timer }}
+            ));
+        }
+    });
+
+    // Make freezing host window, and set its size. The blank page acts as a
     // placeholder so we don't have to repeatedly (and slowly) open and close
     // the window.
-    const windowId = (await browser.windows.create({url: '/pages/blank.html'})).id;
-    const blankTab = (await browser.tabs.query({windowId}))[0];
-    await tabCompletion(blankTab);  // Without this, "Error: No matching message handler"
-    await setViewportSize(blankTab, 1024, 768);
-
-    // Freeze the pages:
-    const lines = document.getElementById('pages').value.split('\n').filter(line => line.length > 0);
-
-    let namesAndUrls;
-    if (lines[0].includes(' ') || lines[0].includes('\t')) {
-        // We have explicit filename prepended to the lines, space-delimited.
-        namesAndUrls = lines.map(function splitAndSuffix(l) {
-            let [name, url] = l.split(/[ \t]+/, 2);
-            return [name + '.html', url];
-        });
-    } else {
-        namesAndUrls = lines.map(l => [undefined, l]);
+    windowId = (await browser.windows.create({
+        url: '/pages/blank.html',
+        width: 1024,
+        height: 768,
+    })).id;
+}
+document.getElementById('freeze').onclick = freezeAllPages;
+
+function fathomNext(event) {
+    const windowId = event.detail;
+
+    // Load next tab, or close the window if we're done.
+    gUrlIndex++;
+    if (gUrlIndex >= gUrls.length) {
+        browser.windows.remove(windowId);
+        document.getElementById('freeze').disabled = false;
+        return;
     }
 
-    const freezeOptions = {wait: parseFloat(document.getElementById('wait').value.trim()),
-                           shouldScroll: document.getElementById('shouldScroll').checked};
-    for (let [filename, url] of namesAndUrls) {
-        try {
-            await freezePage(url, windowId, freezeOptions, filename);
-        } catch (e) {
-            // What can go wrong? Redirects mess up our messaging pipeline.
-            // Modal alerts hang us until the user dismisses them.
-            errorField.appendChild(document.createTextNode(`\nError while freezing ${url}: ${e}`));
+    // Create a new tab with the current url, always as tab[0].
```

Comment:
It seems like the same piece of information is encoded in 2 ways: here by making sure the interesting page is at tab-index 0 and up in freezeAllPages() by testing whether the URL starts with "moz-extension://". Intentional, or did I find a spot where you changed representations partway through but didn't update all references?

Reply:
the process, which this pr doesn't change, is to open the `moz-extension://.../blank.html` file in a new tab, wait for that to load, then repeat(open url in new tab, freeze, close tab).

the `onUpdated` listener needs to wait for the `blank.html` page to finish loading before we can start the collection process, which is why you see the two conditions there.

the old code called `browser.window.create()` then `await tabCompletion()` however `tabCompletion()` has a race condition - if the page completes loading after the initial `isComplete()` check but before the onUpdate listener is attached it'll never see the update event.

i could have created tabs to the right of `blank.html` to have the same effect, but i found always creating into and working on `tab[0]` easier to grok.